File: spark/parsing_xml.py
import pyspark
import pyspark.sql.SqlContext

from xml.sax import ContentHandler, parseString
​
class DistDocHandler(ContentHandler):
    def __init__(self):
        self.lang = "en"
    def startElement(self, tag, attributes):
        if tag == "Article":
            self.lang = attributes["lang"]
​
def extractDoc(distdoc):
    h = DistDocHandler()
    try:
        parseString(distdoc, handler)
    except:
        return "en"
    return handler.lang
​
sc = SparkContext()
sqlContext = SqlContent(sc)
​
sc._jsc.hadoopConfiguration().set("textinputformat.record.delimiter", "<DistDoc")
rdd = sc.textFile("gs://input/*") \
         .map(lambda x: '<DistDoc' + x) \
         .map(lambda x: x.replace('\n', '').replace('\r', '')) \
         .map(lambda x: (extractDoc(x), x))

df = sqlContext.createDataFrame(rdd, ['lang', 'xml'])
df.write.parquet("gs://output/")
​
dfDoc = sqlContext.read.parquet("gs://output=parquet/*")
dfDoc.filter(dfDoc.lang == "en").select(dfDoc.xml) \
    .rdd \
    .saveAsTextFile("gs://output=text/")

File: spark/parsing_xml.scala
import com.databricks.spark.xml.XmlInputFormat
import org.apache.hadoop.io.{LongWritable, Text}
import org.apache.spark._
import scala.xml.XML
import com.databricks.spark.avro._

case class Document(lang: String, xml: String)

def extractAvro(body:String):Document = {
  val doc = XML.loadString(body)
  val docLang = (doc \\ "ArchiveDoc" \ "Article" \ "@lang").headOption.map(x => x.text).getOrElse("")
  Document(docLang, body)
}

sc.hadoopConfiguration.set(XmlInputFormat.START_TAG_KEY, "<DistDoc>")
sc.hadoopConfiguration.set(XmlInputFormat.END_TAG_KEY, "</DistDoc>")

val input = sc.newAPIHadoopFile("gs://xc_snowball_test/201302_0037/file_0001.dsx.gz", classOf[XmlInputFormat], classOf[LongWritable], classOf[Text])

val docs = input.map(x => x._2.toString.filter(_ >= ' ')).map(x => x.replace("<DistDoc>", s"""<DistDoc xmlns="http://xml.dowjones.net/distdoc/v8" xmlns:pd="PrivateData" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">""")).map(extractAvro)
val df = docs.toDF()

df.write.parquet("gs://dj-data-science-datasets/synhub/output=parquet/test1/“)